tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/AutoClientIDShardClusterTest.java

```diff
@@ -0,0 +1,343 @@
+                              if (lastReceived > maxReceived) {
+                                 maxReceived = lastReceived;
+                              }
+                              if (receivedInOrder < 0) {
+                                 receivedInOrder = lastReceived;
+                              } else if (receivedInOrder == lastReceived - 1) {
+                                 receivedInOrder++;
+                              } else {
+                                 if (!orderShot.get()) {
+                                    System.err.println("Sub: " + id + ", received: out of order " + lastReceived + ", last in order: " + receivedInOrder);
+                                 }
+                                 orderShot.set(true);
+                              }
+                           } else {
+                              // no point trying again if there is nothing for us now.
+                              break;
+                           }
+                        }
+                        TimeUnit.MILLISECONDS.sleep(500);
+                     }
+                  }
+               }
+            } catch (Exception ok) {
+            }
+         }
+      }
+   }
+
+   @Ignore("not totally reliable, but does show the root cause of the problem being solved")
+   public void testWithoutOutSharding() throws Exception {
+      setupServers();
+      startServers(0, 1);
+
+      // two bouncy durable consumers
+      DurableSub sub0 = new DurableSub("0");
+      DurableSub sub1 = new DurableSub("1");
+
+      new Thread(sub0).start();
+      new Thread(sub1).start();
+
+      // waiting for registration before production to give bridges a chance
+      assertTrue(sub0.registered.await(20, TimeUnit.SECONDS));
+      assertTrue(sub1.registered.await(20, TimeUnit.SECONDS));
+
+
+      // wait for bindings!
+      assertTrue(Wait.waitFor(() -> 2 == gatherBindingsOnT(0)[0], 20000, 2000));
+
+      assertTrue(Wait.waitFor(() -> 2 == gatherBindingsOnT(1)[0], 20000, 2000));
+
+
+      // wait for remote bindings!
+      assertTrue(Wait.waitFor(() -> 2 == gatherBindingsOnT(0)[1], 20000, 2000));
+
+      assertTrue(Wait.waitFor(() -> 2 == gatherBindingsOnT(1)[1], 20000, 2000));
+
+      // produce a few every second with failover randomize=true to Topic T
+      producer.start();
+
+
+      assertTrue("All sent", Wait.waitFor(() -> toSend.get() == 0));
+
+      assertTrue("All received sub0", Wait.waitFor(() -> sub0.maxReceived == numMessages));
+
+      assertTrue("All received sub1", Wait.waitFor(() -> sub1.maxReceived == numMessages));
+
+
+      // with bouncing, one 'may' be out of order, hence ignored
+      assertTrue(sub0.orderShot.get() || sub1.orderShot.get());
+
+      sub0.consumerDone.set(true);
+      sub1.consumerDone.set(true);
+
+      stopServers(0, 1);
+   }
+
+   @Test
+   public void testWithConsistentHashClientIDModTwo() throws Exception {
```

Comment:
testWithConsistentHashClientIDModTwo is very similar to testWithoutOutSharding, is it possibile reuse the common code?

Reply:
I want to keep them separate but similar b/c the testWithoutOutSharding is non deterministic but informative of the potential problem when producers and consumers bounce around a cluster.